location-ui-component

sirisha-bode opened the pull request:
<!--
    Thank you for your interest in contributing to Nautobot! Please note
    that our contribution policy recommends that a feature request or bug
    report be opened for approval prior to filing a pull request. This
    helps avoid wasting time and effort on something that we might not
    be able to accept.

    Please indicate the relevant feature request or bug report below.
-->
# Closes #<ISSUE NUMBER GOES HERE>
# What's Changed
<!--
    Please include:
    - A summary of the proposed changes
    - A sectioned breakdown for larger features under ## subheadings
-->
# Screenshots
<!--
    - Screenshots, example payloads where relevant:
      - Before/After for bugfixes
      - Using a new feature
-->
# TODO
<!--
    Please feel free to update todos to keep track of your own notes for WIP PRs.
-->
- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/core/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

Sridevi-NTC commented on nautobot/dcim/views.py:
```diff
@@ -255,11 +259,120 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+                return format_html("{}<span>{}</span>", helpers.render_map_button(f"{lat},{lon}"), value)
+            return helpers.HTML_NONE
+
+        if key == "physical_address":
+            if value:
+                return format_html("{}<span>{}</span>", helpers.render_map_button(value), value.replace("\n", "<br>"))
+            return helpers.HTML_NONE
```
Can we use render_address with value transforms?

sirisha-bode replied:
Yes @Sridevi-NTC, updated as requested. We can use render_address with value transforms.

Sridevi-NTC commented on nautobot/dcim/views.py:
```diff
@@ -255,11 +259,120 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+        return super().render_value(key, value, context)
+
+
+class CustomContactInfoPanel(object_detail.ObjectFieldsPanel):
+    def render(self, context):
+        base_content = super().render(context)
+        request = context.get("request")
+        obj = get_obj_from_context(context, self.context_object_key)
+
+        if not request or not obj or not request.user.has_perms(["dcim.contact_association"]):
+            return base_content
+
+        return_url = f"{request.path}?tab=contacts"
+        button_html = helpers.render_contact_team_button(
+            f"{reverse('dcim:location_migrate_data_to_contact', kwargs={'pk': obj.pk})}?return_url={return_url}",
+            "mdi-account-edit",
+            "Convert to contact/team record",
+        )
+
+        if "</div>" in base_content:
+            parts = base_content.rsplit("</div>", 1)
+            return format_html(f"{parts[0]}{button_html}</div>")
+        return base_content
```
Is it necessary to show 'Convert to contact/team record' if contact information is not available?

Sridevi-NTC commented on nautobot/dcim/views.py:
```diff
@@ -268,62 +381,100 @@ class LocationUIViewSet(NautobotUIViewSet):
+                fields=["physical_address", "shipping_address"],
+            ),
+            CustomContactInfoPanel(
+                weight=100,
```
Weight should be unique

Sridevi-NTC commented on nautobot/dcim/views.py:
```diff
@@ -255,11 +259,120 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+        return super().render_value(key, value, context)
+
+
+class CustomContactInfoPanel(object_detail.ObjectFieldsPanel):
+    def render(self, context):
+        base_content = super().render(context)
+        request = context.get("request")
+        obj = get_obj_from_context(context, self.context_object_key)
+
+        if not request or not obj or not request.user.has_perms(["dcim.contact_association"]):
+            return base_content
+
+        return_url = f"{request.path}?tab=contacts"
+        button_html = helpers.render_contact_team_button(
+            f"{reverse('dcim:location_migrate_data_to_contact', kwargs={'pk': obj.pk})}?return_url={return_url}",
+            "mdi-account-edit",
+            "Convert to contact/team record",
+        )
+
+        if "</div>" in base_content:
+            parts = base_content.rsplit("</div>", 1)
+            return format_html(f"{parts[0]}{button_html}</div>")
+        return base_content
```
Doesn't seems correct to me, can this be handled in footer content?

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -268,62 +381,100 @@ class LocationUIViewSet(NautobotUIViewSet):
+                },
+            ),
+            CustomLocationFieldsPanel(
+                weight=100,
```
yes. I've updated the code accordingly.

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -268,62 +381,100 @@ class LocationUIViewSet(NautobotUIViewSet):
-            # show the button if any of these fields have non-empty value.
-            "show_convert_to_contact_button": instance.contact_name or instance.contact_phone or instance.contact_email,
-            **super().get_extra_context(request, instance),
+            "rack_groups": RackGroup.objects.annotate(rack_count=count_related(Rack, "rack_group"))
+            .restrict(request.user, "view")
+            .filter(location__in=related_locations),
+            "stats": {
+                "rack_count": Rack.objects.restrict(request.user, "view").filter(location__in=related_locations).count()
+            },
```
This code provides data for the Rack Groups and Stats panels. It fetches rack groups and rack counts related to the current location and its descendants, while applying user permissions. Without it, those panels wouldn’t display the correct or complete data.

Sridevi-NTC commented on nautobot/core/templatetags/helpers.py:
```diff
@@ -813,6 +815,82 @@ def label_list(value, suffix=""):
+        return format_html("{}", value)
+
+
+def render_map_button(query, icon="mdi-map-marker"):
+    return format_html(
+        '<div class="pull-right noprint">'
+        '<a href="https://maps.google.com/?q={}" target="_blank" class="btn btn-primary btn-xs">'
+        '<i class="mdi {}"></i> Map it</a></div>',
+        query,
+        icon,
+    )
+
+
```
I don't understand how the below code is different from render_address

https://github.com/nautobot/nautobot/blob/c02da6a21e71e8de71b0055392094f3361e2ba23/nautobot/dcim/templates/dcim/location_retrieve.html#L99-L103

https://github.com/nautobot/nautobot/blob/c02da6a21e71e8de71b0055392094f3361e2ba23/nautobot/core/templatetags/helpers.py#L751-L762

sirisha-bode replied:
Thanks for the question! While `render_address()` is well-suited for formatting multi-line textual addresses by splitting them with `<br>` tags and embedding a Google Maps link, it’s not ideal for raw GPS coordinates like latitude and longitude. Coordinates are single-line values (e.g., "17.385, 78.486") and don’t require line breaks or multi-line formatting. Using `render_address()` here would introduce unnecessary formatting and potentially clutter the UI. Instead, `render_map_button()` provides a cleaner and more appropriate "Map it" link without extra formatting, making it better suited for displaying coordinates.

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

Sridevi-NTC commented on nautobot/dcim/views.py:
```diff
@@ -255,11 +259,119 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+
+
+class CustomContactInfoPanel(object_detail.ObjectFieldsPanel):
+    def render(self, context):
+        base_content = super().render(context)
+        request = context.get("request")
+        obj = get_obj_from_context(context, self.context_object_key)
+
+        if not request or not obj or not request.user.has_perms(["dcim.contact_association"]):
+            return base_content
+
+        if not (obj.contact_name or obj.contact_phone or obj.contact_email):
+            return base_content
+
+        return_url = f"{request.path}?tab=contacts"
+        button_html = helpers.render_contact_team_button(
+            f"{reverse('dcim:location_migrate_data_to_contact', kwargs={'pk': obj.pk})}?return_url={return_url}",
+            "mdi-account-edit",
+            "Convert to contact/team record",
+        )
+
+        if "</div>" in base_content:
+            parts = base_content.rsplit("</div>", 1)
+            return format_html(f"{parts[0]}{button_html}</div>")
+
+        return base_content
```
Can we handle this logic in footer section?

itdependsnetworks commented on nautobot/core/templatetags/helpers.py:
```diff
@@ -813,6 +815,61 @@ def label_list(value, suffix=""):
+        return format_html("{}", value)
+
+
+def render_rack_row(indent_px, url, name, count, elevation_url):
```
This shouldn't go into templatetags file. I presume we have come up with a standard, but not sure what that is.

sirisha-bode replied:
Since this is primarily a UI rendering helper and not a filter or tag, I agree it may not belong in templatetags. Could you please clarify the preferred location or standard for such functions? Happy to refactor accordingly

itdependsnetworks replied:
let's just move this to the view.

itdependsnetworks commented on nautobot/dcim/views.py:
```diff
@@ -269,62 +354,104 @@ class LocationUIViewSet(NautobotUIViewSet):
         )
-        stats = {
-            "rack_count": Rack.objects.restrict(request.user, "view").filter(location__in=related_locations).count(),
-            "device_count": Device.objects.restrict(request.user, "view")
```
I see these stats in the screenshot, but can you explain how they are getting there? I would have expected a 1-for-1 mapping here.

itdependsnetworks replied:
leaving this open for core to review.

itdependsnetworks commented on nautobot/dcim/views.py:
```diff
@@ -256,11 +260,92 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+
+
 class LocationUIViewSet(NautobotUIViewSet):
-    # We aren't accessing tree fields anywhere so this is safe (note that `parent` itself is a normal foreign
```
Is this context no longer needed?

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -269,62 +354,104 @@ class LocationUIViewSet(NautobotUIViewSet):
-        if instance is None:
-            return super().get_extra_context(request, instance)
+        if not instance:
+            return {}
```
You're right — the two conditions aren't equivalent. The original check (if instance is None) specifically handles the case when instance is None and ensures we preserve any context from the base class by calling super().

My change (if not instance) could unintentionally match other falsy values and bypass that base logic.

I'll keep the original version to maintain the intended behavior.

sirisha-bode commented:
> From the detail view screenshot I note:
> 
> * Reordering of some fields
> * ~Missing Children~ okay, I see it as a table, not sure if we should keep in left panel as well
> * Children table show difference of nested dots.
> * Both show a line underneath the tab
> * I do not see hierarchy being used, would need to see that still works.

1. Reordering of some fields
Fixed — fields have been reordered as suggested.
Regarding Hierarchy:
Previously, we were seeing the hierarchy because fields="__all__" implicitly included it. However, since I’m now specifying the field order manually, the Hierarchy field is no longer included — it's not an actual model field, so it doesn’t show up unless explicitly handled.
If including the hierarchy is still desired, I can work on adding it back manually (e.g., as a custom display field).
Please confirm if you'd like me to proceed with that.

2. Missing Children
The Children section is already presented as a full-width table. I’ve intentionally removed it from the left panel to avoid duplication. Please confirm if you'd like it shown in both places again — though the full-width panel includes the necessary filtered count, aligned with our refactor across other views.

3. Children table shows difference of nested dots
Handled — I’ve removed the dot-based nesting display as per the comment.

4. Both show a line underneath the tab
I wasn't completely sure what this refers to. Could you please clarify what "both" refers to and which tab the underline appears on? Happy to address once I understand the intent better.

5. I do not see hierarchy being used — would need to see that still works
Currently, there’s no specific hierarchy field in the model or queryset being used here. Since it’s not explicitly defined or utilized in the view, the hierarchy-based behavior isn't triggered. Please let me know if I should integrate a hierarchy-based structure somewhere , or if there’s a specific implementation you’re expecting.

itdependsnetworks commented:
for hierarchy, can you screen shot a nested location?

sirisha-bode commented:
> for hierarchy, can you screen shot a nested location?

<img width="850" height="542" alt="image" src="https://github.com/user-attachments/assets/0bbc8a18-86b5-448a-ad66-98604fdfdf33" />

itdependsnetworks commented on nautobot/dcim/views.py:
```diff
@@ -269,62 +358,102 @@ class LocationUIViewSet(NautobotUIViewSet):
+                    "time_zone",
+                    "description",
+                ],
+                value_transforms={
```
We need to add heirarchy (via ancestors) and then likely update `render_key` on a custom `ObjectFieldsPanel`. 

Then transform, based on this logic"


```html
        {% for ancestor in ancestors %}
            <li>{{ ancestor|hyperlinked_object:"name" }} ({{ ancestor.location_type|hyperlinked_object:"name" }})
                <ul>
        {% endfor %}
        <li><strong>{{ location|hyperlinked_object:"name" }} ({{ location.location_type|hyperlinked_object:"name" }})</strong></li>
        {% for ancestor in ancestors %}
                </ul>
            </li>

```

itdependsnetworks commented on nautobot/extras/tables.py:
```diff
@@ -706,6 +707,44 @@ def log_entry_color_css(record):
+
+    def render_size(self, value, record):
+        try:
+            return f"{record.image.size / 1024:.1f} KB"
```
Actually, there is a https://docs.djangoproject.com/en/5.2/ref/templates/builtins/#filesizeformat already we can use.

itdependsnetworks commented on nautobot/dcim/templates/dcim/location_retrieve.html:
```diff
@@ -1,249 +1,7 @@
-        <div class="panel-heading">
-            <strong>Images</strong>
-        </div>
-        {% include 'inc/image_attachments.html' with images=object.images.all %}
```
We should leave a comment in the `image_attachments.html` to remove once everything is moved out

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

itdependsnetworks commented on nautobot/dcim/views.py:
```diff
@@ -268,6 +269,114 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
 #
 
 
+class LocationFieldsPanel(object_detail.ObjectFieldsPanel):
```
Seems like there should be an easier way, let's sync with Core on this.

sirisha-bode commented on nautobot/core/templatetags/helpers.py:
```diff
@@ -813,6 +815,37 @@ def label_list(value, suffix=""):
 
+@library.filter()
+@register.filter()
+def format_timezone(value):
```
From what I see, format_timezone is meant for rich, human-readable display in the UI (timezone name, UTC offset, and current local time), while tzoffset is a simpler helper that only returns the raw offset (e.g., +0200). They overlap somewhat, but serve different purposes — one for user-friendly output, the other for compact offset-only usage.

sirisha-bode replied:
You’re right — the current code is now using format_timezone instead of the old {{ object.time_zone }} (UTC {{ object.time_zone|tzoffset }}) pattern. Since format_timezone covers both the name and the offset (and even adds local time), it has replaced the need for tzoffset in templates.

Unless we find any remaining references to tzoffset, I think we deprecate or remove it to avoid duplication. That way, we have a single, consistent filter for timezone display across the UI.

HanlinMiao commented on nautobot/dcim/views.py:
```diff
@@ -282,62 +391,103 @@ class LocationUIViewSet(NautobotUIViewSet):
+                label="Geographical Info",
+                fields=["physical_address", "shipping_address"],
+                value_transforms={
+                    "physical_address": [helpers.render_address],
```
missing shipping address

HanlinMiao commented on nautobot/dcim/views.py:
```diff
@@ -282,62 +370,109 @@ class LocationUIViewSet(NautobotUIViewSet):
+                fields=[
+                    "location_type",
+                    "status",
+                    "parent",
```
Probably can do something like
```
def render_key(self, key, value):
     if key == "parent":
          return "Hierarchy"
    return super().render_key(key, value)
```

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -282,62 +370,109 @@ class LocationUIViewSet(NautobotUIViewSet):
+            ),
+        )
+    )
+
```
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode replied:
**Before**
<img width="1380" height="695" alt="image" src="https://github.com/user-attachments/assets/f1afbdcc-0ab0-4303-8657-852a9fdca508" /><img width="902" height="352" alt="image" src="https://github.com/user-attachments/assets/164e599f-775c-4860-9e0a-3f81d6c608dc" />


**After**
<img width="1435" height="702" alt="image" src="https://github.com/user-attachments/assets/35884682-0c88-4ed1-ac67-aa07277e72dc" /> <img width="957" height="353" alt="image" src="https://github.com/user-attachments/assets/49d474f5-05cb-4811-8436-1e246c4e94cd" />

itdependsnetworks commented on nautobot/dcim/views.py:
```diff
@@ -282,62 +370,109 @@ class LocationUIViewSet(NautobotUIViewSet):
+                    "facility",
+                    "asn",
+                    "time_zone",
+                    "description",
```
that's correct.

glennmatthews commented on nautobot/dcim/tables/locations.py:
```diff
@@ -81,3 +82,23 @@ class Meta(BaseTable.Meta):
+
+
+class FlatLocationTable(LocationTable):
+    name = tables.TemplateColumn(
+        template_code=FLAT_LINK,
+        verbose_name="Name",
+        attrs={"td": {"class": "text-nowrap"}},
+    )
```
Perhaps:

```suggestion
    name = tables.Column(linkify=True, attrs={"td": {"class": "text-nowrap"}})
```

sirisha-bode replied:
Thanks for the suggestion! 🙏 I tried switching to
`name = tables.Column(linkify=True, attrs={"td": {"class": "text-nowrap"}})`
but that changes the UI — we lose the tree-style link in the list view. To preserve the existing UI, I think we can keep it as:
```
name = tables.TemplateColumn(
    template_code=TREE_LINK,
    attrs={"td": {"class": "text-nowrap"}},
)
```
or would you prefer me to make the change to the linkify=True version?

glennmatthews commented on nautobot/dcim/tables/locations.py:
```diff
@@ -81,3 +82,23 @@ class Meta(BaseTable.Meta):
         default_columns = ("pk", "name", "status", "parent", "tenant", "description", "tags", "actions")
+
+
+class FlatLocationTable(LocationTable):
```
Do we need this class at all? Or can we just use a `LocationTable(..., hide_hierarchy_ui=True)`?

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -273,6 +276,105 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
 #
 
 
+class RackGroupsFieldsPanel(object_detail.ObjectFieldsPanel):
```
1) Perhaps `LocationRackGroupsPanel`?
2) `ObjectFieldsPanel` seems like the wrong base class here. Should it just be a `Panel` subclass instead?

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -273,6 +276,105 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+            )
+        )
+
+        return format_html(
+            """
+            <div class="panel panel-default">
+                <div class="panel-heading"><strong>Rack Groups</strong></div>
+                <table class="table table-hover panel-body">{}</table>
+            </div>
+            """,
```
Can change this method from a fully custom `render()` to just `render_body_content()` and use the `body_wrapper_template_path="components/panel/body_wrapper_generic_table.html"` added in #7825.

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -273,6 +276,105 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+        )
+
+
+class LocationImagesTablePanel(object_detail.ObjectsTablePanel):
```
Let's see if we can consolidate this with the `DeviceImageAttachmentsTablePanel` added in #7825 and have a single reusable class.

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -273,6 +276,105 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+            return_url += f"?tab={self.tab_id}"
+
+        if self.add_button_route and request.user.has_perms(self.add_permissions or []):
+            try:
+                add_route = reverse(self.add_button_route, kwargs={"object_id": str(obj.pk)})
+                return f"{add_route}?return_url={return_url}"
+            except NoReverseMatch:
+                logger.warning(f"Add route `{self.add_button_route}` could not be reversed with object_id={obj.pk}.")
```
Personally I'd rather remove the try/except and have it bubble up the error normally.

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

glennmatthews commented on nautobot/extras/tables.py:
```diff
@@ -727,6 +728,45 @@ def log_entry_color_css(record):
     return record.log_level.lower()
 
 
+class ImageAttachmentTable(BaseTable):
```
Conflicts with class also added in #7825, please merge them.

sirisha-bode replied:
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode commented on nautobot/core/templatetags/helpers.py:
```diff
@@ -831,6 +831,28 @@ def label_list(value, suffix=""):
+
+    now = datetime.datetime.now(value)
+    result = (
+        f"{value} (UTC {now.strftime('%z')})<br><small>Local time: {now.strftime('%m/%d/%Y %I:%M %p').lower()}</small>"
```
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode commented on nautobot/dcim/models/locations.py:
```diff
@@ -339,3 +339,20 @@ def clean_fields(self, exclude=None):
+        Get image attachments for this Location instance.
+        """
+        # Use the images GenericRelation to fetch related image attachments
+        return self.images.all()
```
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -273,6 +276,105 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+    FIELD_LABEL_OVERRIDES = {
+        "parent": "Hierarchy",
+        "asn": "AS Number",
+    }
```
Updated as requested, should be resolved, let me If I miss anything

itdependsnetworks commented on nautobot/core/templates/inc/image_attachments.html:
```diff
@@ -1,3 +1,4 @@
+{% comment %}3.0 TODO: remove this template, which only exists for backward compatibility with 2.4 and earlier{% endcomment %}
```
```suggestion
{% comment %}3.0 TODO: Potentially remove pending no other references to it. % endcomment %}
```

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -293,30 +392,120 @@ class LocationUIViewSet(NautobotUIViewSet):
+                related_field_name="location",
+                show_table_config_button=False,
+            ),
+            object_detail.ObjectsTablePanel(
+                section=SectionChoices.FULL_WIDTH,
+                weight=100,
+                table_title="Children",
+                table_class=tables.LocationTable,
+                table_attribute="children",
+                related_field_name="parent",
+                order_by_fields=["name"],
+                hide_hierarchy_ui=True,
+            ),
+        )
+    )
+
     def get_extra_context(self, request, instance):
         if instance is None:
             return super().get_extra_context(request, instance)
-        # This query can get really expensive when there are big location trees in the DB. By casting it to a list we
-        # ensure it is only performed once rather than as a subquery for each of the different count stats.
+
+        # This query can get expensive for big trees, so fetch only needed IDs.
         related_locations = list(
             instance.descendants(include_self=True).restrict(request.user, "view").values_list("pk", flat=True)
         )
-        prefix_count_queryset = Prefix.objects.restrict(request.user, "view").filter(locations__in=related_locations)
-        vlan_count_queryset = VLAN.objects.restrict(request.user, "view").filter(locations__in=related_locations)
-        circuit_count_queryset = Circuit.objects.restrict(request.user, "view").filter(
+
+        prefix_qs = Prefix.objects.restrict(request.user, "view").filter(locations__in=related_locations)
+        vlan_qs = VLAN.objects.restrict(request.user, "view").filter(locations__in=related_locations)
+        circuit_qs = Circuit.objects.restrict(request.user, "view").filter(
             circuit_terminations__location__in=related_locations
         )
-        # When there is more than one location, the models that can be assigned to more then one location at the same
-        # time need to be queried with `distinct`. We are avoiding `distinct` when this is not the case, as it incurs
-        # a performance penalty.
+
+        # Avoid distinct() unless multiple locations are involved
```
Why change this?

sirisha-bode commented on nautobot/core/templatetags/helpers.py:
```diff
@@ -831,6 +833,26 @@ def label_list(value, suffix=""):
 
+@library.filter()
+@register.filter()
+def format_timezone(value):
```
Updated as requested, should be resolved, let me If I miss anything

sirisha-bode commented on nautobot/dcim/views.py:
```diff
@@ -278,6 +279,104 @@ class LocationTypeUIViewSet(NautobotUIViewSet):
+
+    def render_value(self, key, value, context):
+        if key == "GPS Coordinates":
+            if value != "Not available":
```
Updated as requested, should be resolved, let me If I miss anything

itdependsnetworks commented on nautobot/core/templates/inc/image_attachments.html:
```diff
@@ -1,3 +1,4 @@
+{% comment %}3.0 TODO: Potentially remove this template pending no other references to it. % endcomment %}
```
```suggestion
{% comment %}3.0 TODO: Potentially remove this template pending no other references to it.{% endcomment %}
```

itdependsnetworks commented on nautobot/core/ui/object_detail.py:
```diff
@@ -1301,9 +1301,6 @@ def render_value(self, key, value, context: Context):
             return display
 
-        if key == "_hierarchy":
-            return render_ancestor_hierarchy(value)
```
@sirisha-bode please ensure to test each of these views for both the queries ran and the UI shown and provide that context here.

sirisha-bode replied:
I’ve checked the difference before and after removing this part, and I can confirm that it does have an impact on RackGroup, TenantGroup, and LocationType. After the removal, the hierarchy is no longer showing for these views. I think we should add it back to restore the hierarchy display.

glennmatthews replied:
Do we just need to add proper hierarchy rendering for the `parent` of those models?

glennmatthews commented on nautobot/dcim/views.py:
```diff
@@ -293,30 +392,120 @@ class LocationUIViewSet(NautobotUIViewSet):
+                },
+                footer_content_template_path="components/panel/footer_convert_to_contact_or_team_record.html",
+            ),
+            object_detail.StatsPanel(
```
I'd prefer a custom `StatsPanel` subclass rather than changing it to an `ObjectFieldsPanel` personally.

itdependsnetworks replied:
@glennmatthews we spoke, and the adjustments were made, was about to resolve this, but should be what you expect at this point.

glennmatthews commented on nautobot/core/ui/object_detail.py:
```diff
@@ -1624,10 +1624,11 @@ def render_body_content(self, context: Context):
-                related_object_count = (
-                    related_object_model_class.objects.restrict(request.user, "view").filter(**filter_dict).count()
-                )
+                filter_dict = {query: list(self.filter_pks)}
```
Minor nit: line 1616 already constructs `self.filter_pks` as a list, I think the explicit cast is only needed at line 1612-1614. Moving the logic there would also avoid re-casting to a list on each iteration through the `self.related_models` loop, though I **think** that's harmless due to queryset caching?

glennmatthews replied:
This change is what is introducing the extra query in the Location detail view, but as I understand it it is necessary at present in order to have the below `.distinct()` not raise an error, which is *probably* a bug in django-tree-queries but also *probably* not worth chasing down at this time.
